src/Confluent.Kafka.Avro/AvroConsumer.cs

```diff
@@ -0,0 +1,201 @@
+        /// </param>
+        public void RegisterAvroDeserializer<T>(AvroDeserializer<T> deserializer)
+        {
+            avroDeserializers.Add(typeof(T), deserializer);
+        }
+
+
+        /// <summary>
+        ///     Removes the avro deserializer associated with the specified type.
+        /// </summary>
+        public void UnregisterAvroSerializer<T>()
+        {
+            avroDeserializers.Remove(typeof(T));
+        }
+
+
+        /// <summary>
+        ///     Gets the avro deserializer that will be used to deserialize values of the specified type.
+        /// </summary>
+        /// <returns>
+        ///     The avro deserializer corresponding to the specified type.
+        /// </returns>
+        public AvroDeserializer<T> GetAvroDeserializer<T>()
+        {
+            return (AvroDeserializer<T>)avroDeserializers[typeof(T)];
+        }
+
+
+        private AvroDeserializer<T> GetOrCreateAvroDeserializer<T>()
+        {
+            if (!avroDeserializers.ContainsKey(typeof(T)))
+            {
+                avroDeserializers.Add(typeof(T), new AvroDeserializer<T>());
+            }
+
+            return (AvroDeserializer<T>)avroDeserializers[typeof(T)];
+        }
+
+
+        /// <summary>
+        ///     Poll for new messages / events. Blocks until a 
+        ///     <see cref="Confluent.Kafka.ConsumeResult{TKey, TValue}" />
+        ///     is available or the operation has been cancelled.
+        /// </summary>
+        /// <param name="keySerdeType">
+        ///     The deserializer type to use to deserialize keys.
+        /// </param>
+        /// <param name="valueSerdeType">
+        ///     The deserializer type to use to deserialize values.
+        /// </param>
+        /// <param name="cancellationToken">
+        ///     A cancellation token that can be used to cancel this operation.
+        /// </param>
+        /// <returns>
+        ///     The <see cref="Confluent.Kafka.ConsumeResult{TKey, TValue}" />.
+        /// </returns>
+        public async Task<ConsumeResult<TKey, TValue>> ConsumeAsync<TKey, TValue>(
+            SerdeType keySerdeType,
+            SerdeType valueSerdeType,
+            CancellationToken cancellationToken = default(CancellationToken))
+        {
+            try
+            {
+                var result = await Task.Run(() => Consume(cancellationToken));
+                if (result == null) { return null; }
+
+                return new ConsumeResult<TKey, TValue>
+                {
+                    TopicPartitionOffset = result.TopicPartitionOffset,
+                    Message = new Message<TKey, TValue>
+                    {
+                        Timestamp = result.Timestamp,
+                        Headers = result.Headers,
+                        Key = keySerdeType == SerdeType.Avro
+                            ? await (GetOrCreateAvroDeserializer<TKey>()).Deserialize(schemaRegistryClient, result.Topic, result.Key, true)
```

Comment:
in sequence. good point, it would be an optimization to make them run in parallel, but only a minor one - they are only blocking on IO when they see a new schema. i've made a note of this